libdrgn/program.c

```diff
@@ -678,6 +678,30 @@ static struct drgn_error *get_prstatus_pid(struct drgn_program *prog, const char
 	return NULL;
 }
 
+static struct drgn_error *get_prstatus_pgrp(struct drgn_program *prog, const char *data,
+					   size_t size, uint32_t *ret)
```

Comment:
`pgrp` isn't quite what we want here. That's the process group ID: see [Wikipedia](https://en.wikipedia.org/wiki/Process_group) and [credentials(7)](https://man7.org/linux/man-pages/man7/credentials.7.html). Process groups are usually used for shell pipelines. For example, if I do:

```
$ python3 -c 'import os, sys; print("Pipe PID:", os.getpid(), file=sys.stderr)' | ./crashme_multithreaded
Main TID: 3886869
Second TID: 3886870
Pipe PID: 3886868
zsh: segmentation fault (core dumped)  ./crashme_multithreaded
```

Then `pgrp` is the PID of the first process in the pipeline:

```
$ eu-readelf -n core.3886869 | grep PRSTATUS -A 4
  CORE                 336  PRSTATUS
    info.si_signo: 11, info.si_code: 0, info.si_errno: 0, cursig: 11
    sigpend: <>
    sighold: <>
    pid: 3886870, ppid: 3868213, pgrp: 3886868, sid: 3868213
--
  CORE                 336  PRSTATUS
    info.si_signo: 11, info.si_code: 0, info.si_errno: 0, cursig: 11
    sigpend: <>
    sighold: <>
    pid: 3886869, ppid: 3868213, pgrp: 3886868, sid: 3868213
```

So we really do need to get the main TID from the `pid` field of `PRPSINFO` as described in #144:

```
$ eu-readelf -n core.3886869 | grep PRPSINFO -A 2
  CORE                 136  PRPSINFO
    state: 2, sname: D, zomb: 0, nice: 0, flag: 0x0000000040400404
    uid: 66107, gid: 100, pid: 3886869, ppid: 3868213, pgrp: 3886868
```

You should be able to do that in the same loop over notes in `drgn_program_cache_core_dump_notes()`. Note that you can ignore `drgn_program_cache_kdump_notes()`, as that's only used for kernel core dumps.

Reply:
> Thanks for taking this! The scaffolding all looks right here, but I commented on where we should be getting the main TID from.

Thank for super quick review! Going to fix everything